---
title: "Making a custom database - ncbi Example"
author: "Salix Dubois"
date: "30/09/2020"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, collapse = TRUE)
```
## Prep
These packages will be needed.
The code of the custom functions used for this is here : https://github.com/salix-d/customSeqDB
Functions specific to this db are here : https://github.com/salix-d/customSeqDB/blob/master/R/ncbi-db.R
```{r prep, eval=FALSE}
#I'm not sure this one would install even if it's a dependency because it's done through BiocManager
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("ShortRead")

#This one should, but can still install it prior.
install.packages("jsonlite")


if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("salix-d/customSeqDB")
library(customSeqDB)
```


## Searching for records and getting their details
This gets the ESummary of all records in the NCBI 'nucleotide' database matching these taxa and gene which have the word 'complete' in the 'title' fields but don't have 'UNVERIFIED' in the 'keyword' fields. The query isn't case sensitive. For the usable fields and their name see : https://www.ncbi.nlm.nih.gov/books/NBK49540/

It then parse the data and returns a list with history server information in case we want to look at the records again and a list of details including the accession numbers, the taxonomic ids and the description. The ESearch can return up to 100k results. This part may take a while.

```{r details}
taxa <- c("Opilioacarida", "Holothyrida", "Ixodida", "Mesostigmata", "Trombidiformes", "Sarcoptiformes")
gene <- c("COI", "CO1", "COXI", "COX1") # ncbi includes 6 orders and the COI gene can be written 4 ways
ncbi <- get_ncbiDetails(taxa = taxa, gene = gene, customTerm = "AND complete [TITL] NOT unverfified[KYWD]")
head(ncbi$details)
```
To see how many records were found for each taxon, the queries can be separated. Using the history server information that it returns, they can all be added to the same WebEnv. See https://www.ncbi.nlm.nih.gov/books/NBK25499/#_chapter4_ESummary_ for details.
```{r detailsByTaxa, eval=FALSE}
ncbi <- list(Opilioacarida = get_ncbiDetails("Opilioacarida", c("COI", "CO1", "COXI", "COX1"), customTerm = "AND complete[ALLl] NOT UNVERIFIED[KYWD]"))
for(o in c("Holothyrida", "Ixodida", "Mesostigmata", "Trombidiformes", "Sarcoptiformes")){
  ncbi[[o]] <- get_ncbiDetails(o, c("COI", "CO1", "COXI", "COX1"), customTerm = "AND complete[ALLl] NOT UNVERIFIED[KYWD]", query = ncbi$Opilioacarida$query)
}
#ncbi$details <- do.call(rbind, lapply(ncbi[1:6], function(x) x$details))
sapply(ncbi$details, ncol)
ncbi$Opilioacarida$details <- t(ncbi$Opilioacarida$details) #(if only one result came back it needs to be transpose to be merged wih the rest.)
ncbi$Holothyrida$details <- t(ncbi$Holothyrida$details)
ncbi$details <- NULL
ncbi$details <- do.call(rbind, lapply(ncbi[1:6], function(x) x$details))
```
<i>Sometimes it return 'Error in parse_con(txt, bigint_as_char) : parse error: premature EOF', but rerunning fixes it... I'm not sure why.</i>

## Filtering the details to get only "complete" COI
<i>This step will need to be adapted depending on your query.</i>
The fltr string looks for the word 'complete' a maximum of two commas after the gene (because farther than that it's not the COI gene that is complete; based on these descriptions). Some sequences won't have the word for the gene but will be complete genome/sequence and can have a complete gene, so they are included. The gene is looked for with COx?[I1] because sometimes there's an X but not always and the one can be arabic or roman. 'ignore.case = T' because the gene can be in all caps or not.
<i>some sequences will still be partial sequences and so, they will be filtered again once we have the sequences</i>
```{r filterDetails}
nrow(ncbi$details)
fltr <- "COX?[I1]\\)?[[:alnum:] \\(\\);]*,?[[:alnum:] \\(\\);]*, complete|complete genome|complete sequence"
ncbi$complete$details <- ncbi$details[grep(fltr, ncbi$details$description, ignore.case = T),]
nrow(ncbi$complete$details)
```
## Getting the "complete" sequences files
This uses EFetch to fetch the records for all provided accession numbers in the specified format and parse them to return the accession number, the taxid and the region of the sequence for the specified gene(s).
Since CDS have reading frame, 'codon_start' is set to true to adjust the beginning of the sequence according to start position of the start codon.

If the difference between the number of records returned by the search and the number of recoreds filtered isn't big, it's much faster to use the history server to fetch them.
When using ids to fetch, it can only fetch 200 at a time, with the server it can fetch up to 10k at a time.
```{r fetchWithAC}
ncbi$complete$filesByAcc <- fetch_ncbiSeq(accList = ncbi$complete$details$accession, gene = "COX?[I1]\"", gene_type = "CDS", codon_start = T, saveRec = T)
ncbiFiles <- sapply(c("002", "003", "004"), function(i) gsub("001", i, ncbiFile))
ncbi$complete$filesByAcc <- do.call(rbind, lapply(ncbiFiles, function(file) parse_flatFile(file, gene = "COX?[I1]\"", gene_type = "CDS", codon_start = T)))
nrow(ncbi$complete$filesByAcc)
head(ncbi$complete$filesByAcc)
```

```{r fetchWithQ}
ncbi$complete$files <- fetch_ncbiSeq(query = ncbi$query, gene = "COX?[I1]\"", gene_type = "CDS", codon_start = T, saveRec = T)
nrow(ncbi$complete$files)
head(ncbi$complete$files)
```

If the file has already been downloaded, it can be reparsed with :
```{r parseFromFile_show, eval=FALSE}
ncbiFile <- "path/to/file"
```

```{r parseFromFile, echo=TRUE}
ncbi$complete$files <- parse_flatFile(ncbiFile, gene = "COX?[I1]\"", gene_type = "CDS", codon_start = T)
nrow(ncbi$complete$files)
head(ncbi$complete$files)
```
The number of rows might be smaller than the number of record we fetched :
- when records are part of a 'popset', only one record is returned. To get all of them, the missing accession can be refetched.
- some records may not have had the correct information and were ignored. (there will be a warning. They should be remove from the missing acc to be refetched, or else they will be parse for nothing and be ignored again)

*If you don't want all records of the popsets, skip to next step (filtering description). If you are going to merge this db with another one, I would keep track of the popsets' accession number. Other db might give you the whole set, so you could remove the ones you don't want to keep.*
```{r}
ncbi$complete$details$accession <- gsub("\\.\\d*$", "", ncbi$complete$details$accession)
tmp.accList <- ncbi$complete$details$accession[!ncbi$complete$details$accession %in% ncbi$complete$files$accession]
I <- length(tmp.accList)
i <- 0
tmp.accList <- tmp.accList[!tmp.accList %in% c(tmp$accession, "AH011506")]
while(length(tmp.accList)!=0){
  tmp.accList <- tmp.accList[!tmp.accList %in% tmp$accession]
  tmp <- rbind(tmp, fetch_ncbiSeq(accList = tmp.accList, gene = "COX?[I1]\"", gene_type = "CDS", codon_start = T, saveRec = T))
  i <- i + 1
  if(i>I) break 
  #bc even if fetchingone by one it shouldn't reach that. this mean there's an acc that just doesn't return anything
}
 
sum(duplicated(tmp$sequence))
```
In 78 records of 2 popset, only 7 sequences are duplicates.

They can be added to the main data.frame, duplicates will be filtered out in next steps.
```{r}
ncbi$complete$files <- rbind(ncbi$complete$files, tmp)
rownames <- NULL
nrow(ncbi$complete$files)
```

## Filtering the description for complete *COI*
```{r}
grep(fltr, tmp$description, value = T, invert = T)
fltr <- "COX?[I1]\\)?[[:alnum:]\\s\\(\\);]*,?[[:alnum:]\\s\\(\\);]*,\\s*complete|complete\\s*genome|complete\\s*sequence|CoI\\)\\s*gene,\\s*complete"
#Changing name to do modification and keep a backup
ncbi$complete$COI <- ncbi$complete$files[grepl(fltr, ncbi$complete$files$description),]
rownames(ncbi$complete$COI) <- NULL
nrow(ncbi$complete$COI)
```
## Removing the duplicated records
Some sequences will be duplicated.
```{r}
sum(duplicated(ncbi$complete$COI$sequence))
```
If they are from different taxId, they will be kept, if not, they will be remove.
```{r}
seqByTax <- lapply(unique(ncbi$complete$COI$taxId), function(id) ncbi$complete$COI[ncbi$complete$COI$taxId == id,])
dupSeq <- sapply(seqByTax, function(x) rownames(x)[duplicated(x$sequence)])
as.data.frame(rbind(n.seqByTax = sapply(seqByTax, nrow), n.dupSeq = sapply(dupSeq, length)))
length(unlist(dupSeq))
ncbi$complete$COI <- ncbi$complete$COI[-as.numeric(unlist(dupSeq)),]
rownames(ncbi$complete$COI) <- NULL
nrow(ncbi$complete$COI)
```
## Filtering the "complete" sequences fasta files
### Fixing translation exception
Before this can be done, if there are translation exceptions, they should be checked and fixed.
```{r}
sum(!is.na(ncbi$complete$COI$transl_except))
```
There are 4 in this case.
Start transl_except will have 'Met' and end ones will have 'TERM', so they can be found with grep.
```{r}
ncbi$complete$COI[grep("Met", ncbi$complete$COI$transl_except), -c(2:6)]
```
For the first 2, has.start is already set to T because the AA at the range specified by the exception already is a Met; not sure why this is there.
The 3rd one is 'ANT' but should be 'ATT', so it will be modified.
```{r}
ncbi$complete$COI$sequence[grep("Met", ncbi$complete$COI$transl_except)[3]] <- gsub("^ANT", "ATT", ncbi$complete$COI$sequence[grep("Met", ncbi$complete$COI$transl_except)[3]])
ncbi$complete$COI[grep("Met", ncbi$complete$COI$transl_except), -c(2:6)]
```
Starts are fixed. Now the ends.
```{r}
ncbi$complete$COI[grep("TERM", ncbi$complete$COI$transl_except), -c(2:6)]
```
This one is actually missing 2 nucleotides and they should be added.

*In doubts, the line below confirms that there is one extra nucleotide.*
```{r}
nchar(ncbi$complete$COI[grep("TERM", ncbi$complete$COI$transl_except), 12])%%3
```
So the two 'A' are gonna be added.
```{r}
ncbi$complete$COI$sequence[grep("TERM", ncbi$complete$COI$transl_except)] <- paste0(ncbi$complete$COI$sequence[grep("TERM", ncbi$complete$COI$transl_except)], "AA")
ncbi$complete$COI[grep("TERM", ncbi$complete$COI$transl_except), -c(2:6)]
```

### Checking if all the sequences have a start and an end codons like they should.

The start and end codons vary depending on the taxa. See the [translation tables](https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi)
*This example uses the Invertabrate Mitochondrion Code (transl_table5).*
```{r}
start_codons <- c("TTG","ATT","ATC","ATA","ATG","GTG")
end_codons <- c("TAA", "TAG")

has.start <- grep(paste0("^", paste(start_codons, collapse = "|^")), ncbi$complete$COI$sequence, ignore.case = T)
has.end <- grep(paste0(paste(end_codons, collapse = "$|"), "$"), ncbi$complete$COI$sequence, ignore.case = T)
ncbi$complete$COI$has.start = F
ncbi$complete$COI$has.start[has.start] = T
ncbi$complete$COI$has.end = F
ncbi$complete$COI$has.end[has.end] = T
ncbi$complete$COI$complete = F
ncbi$complete$COI$complete[ncbi$complete$COI$has.start == T & ncbi$complete$COI$has.end == T] = T
sum(!ncbi$complete$COI$complete)
```
But some are gonna be complement sequence, so start and end codons must be looked for on it's reversed complement.
```{r}
sum(ncbi$complete$COI$complement)
```

```{r}
comp <- list(A = "T", C = "G", "T" = "A", G = "C")
start_codonsRC <- sapply(start_codons, function(x) paste(rev(comp[strsplit(x, "")[[1]]]), collapse = ""))
ncbi$complete$COI$has.start[ncbi$complete$COI$complement] <- grepl(paste0(paste(start_codonsRC, collapse = "$|"), "$"), ncbi$complete$COI$sequence[ncbi$complete$COI$complement])

end_codonsRC <- sapply(end_codons, function(x) paste(rev(comp[strsplit(x, "")[[1]]]), collapse = ""))
ncbi$complete$COI$has.end[ncbi$complete$COI$complement] <- grepl(paste0("^", paste(end_codonsRC, collapse = "|^")), ncbi$complete$COI$sequence[ncbi$complete$COI$complement])

ncbi$complete$COI$complete[ncbi$complete$COI$has.start == T & ncbi$complete$COI$has.end == T] = T
sum(!ncbi$complete$COI$complete)
```

```{r}
incompletes <- which(!ncbi$complete$COI$complete)
length(incompletes)
ncbi$complete$COI[incompletes,-c(2:6,12)] #to look at transl_except and notes
```
Some have undetermined start codons. Not sure what to do with these. Would probably check records manually to see how much is missing. 
Some are because the frame was missing, could look at records/see if reading it from frame 2/3 would fix.
The rest should be look at individually, most will be 'complete genome' but with gab before/after the gene region. If the gab length is known, They could still be added.

Making a new data.frame 'checked' with all complete sequences. If the rest is fixed, they can be added.
```{r}
ncbi$complete$checked <- ncbi$complete$COI[ncbi$complete$COI$complete,]
nrow(ncbi$complete$checked)
```

*The next steps can be skipped partially or entirely depending on the time that can be put on this.*
The sequences with issues will be separated in different data.frames so they can be check and added to the 'checked' one if they are fixed.
```{r}
isNotDetermined <- grep("start codon not determined", ncbi$complete$COI$note)
ncbi$complete$NotDetermined <- ncbi$complete$COI[isNotDetermined,] 

isMissingFrame <- grep("frame", ncbi$complete$COI$note)
ncbi$complete$missingFrame <- ncbi$complete$COI[isMissingFrame,]
ncbi$complete$OtherIncomplete <- ncbi$complete$COI[incompletes,][!incompletes %in% c(isNotDetermined, isMissingFrame),]
sapply(c("NotDetermined","missingFrame","OtherIncomplete"), function(x) nrow(ncbi$complete[[x]]))
```
**missing reading frame field**
```{r}
nchar(ncbi$complete$missingFrame$sequence) %% 3
```
Since this returns 0 and the sequence has a start codon, the reading frame is probably right.

If we look at other squences for the same taxId...
```{r}
ncbi$complete$missingFrame$taxId %in% ncbi$complete$COI$taxId
isSameId <- which(ncbi$complete$COI$taxId %in% ncbi$complete$missingFrame$taxId)
ncbi$complete$COI[isSameId,]
sapply(ncbi$complete$COI$sequence[isSameId], nchar, USE.NAMES = F)
```
All the others are complete, except the first one. When checking their length, they are, respectively, shorter and longer than the rest.

When looking at the records, there's no gene between the end of COX1 and the beginning of COX2, if we add it to the sequence and translate it again, there is an end codon. And the 27 nucleotide before it matches the 27 nuclotides prior to the stop codon of all other complete sequences (the RC (11) one wasn't checked) of the same taxId.
```{r}
z <- ncbi$complete$COI$sequence[isSameId[-c(1,10:11)]]
all(sapply(z, substr, 1521, 1551, USE.NAMES = F) %in% substr(z[1], 1521, 1551))
tmp.seq <- paste0(ncbi$complete$missingFrame$sequence, "AATTTAGAAAGGTTTAAAATTTTTTTAAATTATACAAAATATTAATAAAATAAAATAAA")
tmp.seq.tranl <- actg2imc(tmp.seq)
grep("\\*", tmp.seq.tranl)
grep("\\*", strsplit(tmp.seq.tranl, "")[[1]])
nchar(tmp.seq.tranl)
#it's 5th from the end, so removing 4*3(12) nucleotides.
nchar(tmp.seq)%%3
# 2 extra nucleotide at the end, so removing 14 total
tmp.seq <- substr(tmp.seq, 1, nchar(tmp.seq)-14)
nchar(tmp.seq)
substr(tmp.seq, 1539, 1569) == substr(z[1], 1521, 1551)
```
Uncertain what to do with this info though...

For the rest it would be manual check up.

## Getting the taxonomy for the remaining sequences
### Making the taxonomy sql table
The function will download a taxdump.tar.gz file from NCBI servers and extract the names.dmp and nodes.dmp files from it.
Then thes files will be merged into one tax.dmp file keeping only the columns required (id, name, rank and parentId)
For names.dmp, also only keeps the rows containing the scientific names so there's only one name per id.
It will then use that file to create a SQLite database containing a taxonomy table from which we can get taxonomic information from taxIds with the get_taxonomy() function.
get_taxonomy gets the the parentId of all parents of each taxIds and returns the names corresponding to the ranks desired a data frame with the taxids as row names and ranks as columns.
```{r}
db_path <- mk_taxonomy.table(dbFile = paste0('./ncbi-db_', Sys.Date()))
ncbi$complete$tax <- get_taxonomy(ncbi$complete$checked$taxId, sqlFile = db_path, ranks = c("kingdom","phylum","class","order","family","genus","species"))
#to see what taxa the db covers :
apply(ncbi$complete$tax[4:5], 2, table)
```

## Writing the dada2 formated fasta file
```{r}
ncbi$complete$tax$id <- paste0(apply(ncbi$complete$tax[1:7], 1, paste0, collapse = ";"), ";")
write_fasta(sequences = ncbi$complete$checked$sequence, ids = ncbi$complete$tax$id, outFile = "completeCOI_ncbi_ff_267seq.fasta")
```

## Testing the database
```{r}
seqtab <- as.matrix(read.csv("../../MitesDB/data/seqtab.nochim_MITES.csv", row.names = 1))
library(dada2)
set.seed(20) #this doesn't even work... 
ncbi$taxa <- dada2::assignTaxonomy(seqtab, "./completeCOI_ncbi_ff.fasta", tryRC = T, outputBootstraps = T)
#sapply(names(ncbi$taxa), function(name) write.csv(ncbi$taxa[[name]], paste0("completeCOI_ncbi_ff_", name, ".csv")))
ncbi$taxa.p = as.data.frame(ncbi$taxa$tax, row.names = seq_len(nrow(ncbi$taxa$tax)))
ncbi$boot.p = as.data.frame(ncbi$taxa$boot, row.names = seq_len(nrow(ncbi$taxa$boot)))

ncbi$taxa2 <- dada2::assignTaxonomy(seqtab, "completeCOI_ncbi_ff_267seq.fasta", tryRC = T, outputBootstraps = T)
#sapply(names(ncbi$taxa2), function(name) write.csv(ncbi$taxa2[[name]], paste0("completeCOI_ncbi_ff_267seq_", name, ".csv")))
ncbi$taxa2.p = as.data.frame(ncbi$taxa2$tax, row.names = seq_len(nrow(ncbi$taxa2$tax)))
ncbi$boot2.p = as.data.frame(ncbi$taxa2$boot, row.names = seq_len(nrow(ncbi$taxa2$boot)))
# To see how many sequences are assigned per ranks
colSums(!is.na(ncbi$taxa.p))
colSums(!is.na(ncbi$taxa2.p))

sum(is.na(ncbi$taxa.p$Family) != is.na(ncbi$taxa2.p$Family))
#same number of families but not the same sequences
orderNAs <- which(is.na(ncbi$taxa2.p$Order))[which(!which(is.na(ncbi$taxa2.p$Order)) %in% which(is.na(ncbi$taxa.p$Order)))]
cbind(
  ncbi$taxa.p$Order[orderNAs],
  ncbi$boot.p$Order[orderNAs],
  ncbi$taxa2.p$Order[orderNAs],
  ncbi$boot2.p$Order[orderNAs]
  )[c(1,3,26,27,2,4:25,28),]
#the boot confidences doesn't change that much for the order rep by only one sequence, but it changes for the one that got the popset sequences.
```
